File: src/ios/BlinkUpPlugin.m
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Created by Stuart Douglas (sdouglas@macadamian.com) on June 11, 2015.
 * Copyright (c) 2015 Macadamian. All rights reserved.
 */

#import "BlinkUpPlugin.h"
#import "BlinkUpPluginResult.h"
#import <BlinkUp/BlinkUp.h>

NSString * const PLAN_ID_CACHE_KEY = @"planId";

// status codes
typedef NS_ENUM(NSInteger, BlinkUpStatusCodes) {
    DEVICE_CONNECTED              = 0,
    GATHERING_INFO                = 200,
    CLEAR_WIFI_COMPLETE           = 201,
    CLEAR_WIFI_AND_CACHE_COMPLETE = 202
};

// error codes
typedef NS_ENUM(NSInteger, BlinkUpErrorCodes) {
    INVALID_ARGUMENTS   = 100,
    PROCESS_TIMED_OUT   = 101,
    CANCELLED_BY_USER   = 102,
    INVALID_API_KEY     = 103
};

typedef NS_ENUM(NSInteger, StartBlinkupArguments) {
    StartBlinkUpArgumentApiKey = 0,
    StartBlinkUpArgumentDeveloperPlanId,
    StartBlinkUpArgumentIsInDevelopment,
    StartBlinkUpArgumentTimeOut,
};

@implementation BlinkUpPlugin

/*********************************************************
 * Parses arguments from javascript and displays BlinkUp
 ********************************************************/
- (void)startBlinkUp:(CDVInvokedUrlCommand*)command {
    NSLog(@"startBlinkUp Started.");

    _callbackId = command.callbackId;

    [self.commandDelegate runInBackground:^{
        _apiKey = [command.arguments objectAtIndex:StartBlinkUpArgumentApiKey];
        _developerPlanId = [command.arguments objectAtIndex:StartBlinkUpArgumentDeveloperPlanId];
        _isInDevelopment = [[command.arguments objectAtIndex:StartBlinkUpArgumentIsInDevelopment] boolValue];
        _timeoutMs = [[command.arguments objectAtIndex:StartBlinkUpArgumentTimeOut] integerValue];

        if ([self sendErrorToCallbackIfArgumentsInvalid]) {
            return;
        }

        NSLog(@"startBlinkUp. isInDevelopment? %d, timeoutMs? %ld", _isInDevelopment, (long)_timeoutMs);

        [self navigateToBlinkUpView];
    }];
}

/*********************************************************
 * Cancels device polling
 ********************************************************/
- (void)abortBlinkUp:(CDVInvokedUrlCommand *)command {
    NSLog(@"abortBlinkUp Started.");

    _callbackId = command.callbackId;

    [self.commandDelegate runInBackground:^{
        [_blinkUpController.devicePoller stopPolling];

        BlinkUpPluginResult *abortResult = [[BlinkUpPluginResult alloc] init];
        abortResult.state = Error;
        [abortResult setPluginError:CANCELLED_BY_USER];

        [self sendResultToCallback:abortResult];
    }];
}

/********************************************************
 * Clears wifi configuration of Imp and cached planId
 ********************************************************/
- (void) clearBlinkUpData:(CDVInvokedUrlCommand *)command {
    NSLog(@"clearBlinkUpData Started.");

    _callbackId = command.callbackId;

    [self.commandDelegate runInBackground:^{
        // clear cached planId
        [[NSUserDefaults standardUserDefaults] setObject:nil forKey:PLAN_ID_CACHE_KEY];

        // create a controller to clear network info
        BUNetworkConfig *clearConfig = [BUNetworkConfig clearNetworkConfig];

        if(_flashController == nil) {
            _flashController = [[BUFlashController alloc] init];
        } else if(_blinkUpController != nil) {
            _flashController = _blinkUpController.flashController;
        }

        dispatch_async(dispatch_get_main_queue(), ^{
            // present the clear device flashing screen
            [_flashController presentFlashWithNetworkConfig:clearConfig configId:nil animated:YES resignActive:
             ^(BOOL willRespond, BUDevicePoller *devicePoller, NSError *error) {
                 [self blinkUpDidComplete:false userDidCancel:false error:nil clearedCache:true];
             }];
        });
    }];
}


/*********************************************************
 * shows default UI for BlinkUp process. Modify this method
 * if you wish to use a custom UI (refer to API docs)
 ********************************************************/
- (void) navigateToBlinkUpView {
    // Uses development plan Id if our device is still in development.
    // Alternatively, tries to load a cached planID that would be generated for a blessed device by electric imp.
    // IMPORTANT NOTE: if a developer planId makes it into production, the device will NOT connect.
    // See electricimp.com/docs/manufacturing/planids/ for more info about planIDs
    NSString *planId = _isInDevelopment
                            ? _developerPlanId
                            : [[NSUserDefaults standardUserDefaults] objectForKey:PLAN_ID_CACHE_KEY];

#ifdef DEBUG
    planId = _developerPlanId;
#endif

    if (planId == _developerPlanId) {
        NSLog(@"WARNING - Using Developer Plan. For production, set isInDevelopment flag to false.");
    }

    if (_blinkUpController == nil) {
        _blinkUpController = (planId == nil)
                                ? [[BUBasicController alloc] initWithApiKey:_apiKey]
                                : [[BUBasicController alloc] initWithApiKey:_apiKey planId:planId];
    }

    dispatch_async(dispatch_get_main_queue(), ^{
        [_blinkUpController presentInterfaceAnimated:YES
            resignActive: ^(BOOL willRespond, BOOL userDidCancel, NSError *error) {
                [self blinkUpDidComplete:willRespond userDidCancel:userDidCancel error:error clearedCache:false];
            }
            devicePollingDidComplete: ^(BUDeviceInfo *deviceInfo, BOOL timedOut, NSError *error) {
                [self deviceRequestDidCompleteWithDeviceInfo:deviceInfo timedOut:timedOut error:error];
            }
        ];
    });
}


/*********************************************************
 * Called when BlinkUp controller is closed, by user
 * cancelling, flashing process complete, or on error.
 * Sends status back to Cordova app.
 ********************************************************/
- (void) blinkUpDidComplete:(BOOL)willRespond userDidCancel:(BOOL)userDidCancel error:(NSError*)error clearedCache:(BOOL)clearedCache {

    NSLog(@"blinkUpDidComplete Started. willRespond: %d userDidCancel: %d, Error: %@",
          willRespond, userDidCancel, error);

    BlinkUpPluginResult *pluginResult = [[BlinkUpPluginResult alloc] init];

    if (willRespond) {
        pluginResult.state = Started;
        pluginResult.statusCode = GATHERING_INFO;
    }
    else if (userDidCancel) {
        pluginResult.state = Error;
        [pluginResult setPluginError:CANCELLED_BY_USER];
    }
    else if (error != nil) {
        pluginResult.state = Error;
        [pluginResult setBlinkUpError: error];
    }
    else {
        pluginResult.state = Completed;
        if (clearedCache) {
            pluginResult.statusCode = CLEAR_WIFI_AND_CACHE_COMPLETE;
        }
        else {
            pluginResult.statusCode = CLEAR_WIFI_COMPLETE;
        }
    }

    [self sendResultToCallback:pluginResult];
}


/*********************************************************
 * Called when device info has been loaded from Electric
 * Imp server, or when that request timed out.
 * Sends device info and status back to Cordova app.
 ********************************************************/
- (void) deviceRequestDidCompleteWithDeviceInfo:(BUDeviceInfo*)deviceInfo timedOut:(BOOL)timedOut error:(NSError*)error {

    NSLog(@"deviceRequestDidComplete Started. DeviceInfo: %@ TimedOut?: %d, Error: %@",
          deviceInfo, timedOut, error);

    BlinkUpPluginResult *pluginResult = [[BlinkUpPluginResult alloc] init];

    if (timedOut) {
        pluginResult.state = Error;
        [pluginResult setPluginError:PROCESS_TIMED_OUT];
    }
    else if (error != nil) {
        pluginResult.state = Error;
        [pluginResult setBlinkUpError:error];
    }
    else {
        // cache plan ID if it's not development ID (see electricimp.com/docs/manufacturing/planids/)
        if (![deviceInfo.planId isEqual: _developerPlanId]) {
            [[NSUserDefaults standardUserDefaults] setObject:deviceInfo.planId forKey:PLAN_ID_CACHE_KEY];
        }

        pluginResult.state = Completed;
        pluginResult.statusCode = DEVICE_CONNECTED;
        pluginResult.deviceInfo = deviceInfo;
    }

    [self sendResultToCallback:pluginResult];
}

/*********************************************************
 * Sends error to callback if arguments don't have correct
 * type, or if apiKey is invalid format.
 * @return YES if error was sent, NO otherwise
 ********************************************************/
- (BOOL) sendErrorToCallbackIfArgumentsInvalid {

    BOOL invalidArguments = self.timeoutMs == 0 || [self.developerPlanId length] == 0;
    BOOL invalidApiKey = ![BlinkUpPlugin isApiKeyFormatValid:self.apiKey];

    // send error to callback
    if (invalidArguments || invalidApiKey) {
        BlinkUpPluginResult *pluginResult = [[BlinkUpPluginResult alloc] init];
        pluginResult.state = Error;

        [pluginResult setPluginError:(invalidApiKey ? INVALID_API_KEY : INVALID_ARGUMENTS)];

        [self sendResultToCallback:pluginResult];
    }

    return (invalidArguments || invalidApiKey);
}

/*********************************************************
 * Creates a cordova plugin result from pluginResult with
 * correct settings and sends to callback
 ********************************************************/
- (void) sendResultToCallback:(BlinkUpPluginResult *)pluginResult {
    CDVPluginResult *cordovaResult = [CDVPluginResult resultWithStatus:[pluginResult getCordovaStatus] messageAsString:[pluginResult getResultsAsJsonString]];
    [cordovaResult setKeepCallbackAsBool: [pluginResult getKeepCallback]];
    [self.commandDelegate sendPluginResult:cordovaResult callbackId:_callbackId];
}

/*********************************************************
 * Returns true iff api key is 32 alphanumeric characters
 ********************************************************/
+ (BOOL) isApiKeyFormatValid: (NSString *)apiKey {
    if (apiKey == nil || apiKey.length != 32) {
        return NO;
    }

    // must be only alphanumeric characters
    NSCharacterSet *alphaSet = [NSCharacterSet alphanumericCharacterSet];
    return ([[apiKey stringByTrimmingCharactersInSet:alphaSet] length] == 0);
}

@end

File: src/android/BlinkUpPlugin.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Created by Stuart Douglas (sdouglas@macadamian.com) on June 11, 2015.
 * Copyright (c) 2015 Macadamian. All rights reserved.
 */

package com.macadamian.blinkup;

import android.app.Activity;
import android.content.Intent;
import android.text.TextUtils;
import android.util.Log;

import com.electricimp.blinkup.BlinkupController;
import com.electricimp.blinkup.TokenStatusCallback;
import com.electricimp.blinkup.TokenAcquireCallback;
import com.electricimp.blinkup.ServerErrorHandler;
import com.macadamian.blinkup.util.PreferencesHelper;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONArray;
import org.json.JSONException;

/*********************************************
 * execute() called from Javascript interface,
 * which saves the arguments and presents the
 * BlinkUpPlugin interface from the SDK
 ********************************************/
public class BlinkUpPlugin extends CordovaPlugin {
    private static final String TAG = "BlinkUpPlugin";

    private static final String START_BLINKUP = "startBlinkUp";
    private static final String ABORT_BLINKUP = "abortBlinkUp";
    private static final String CLEAR_BLINKUP_DATA = "clearBlinkUpData";

    private static CallbackContext sCallbackContext;
    private static boolean sClearCache = false;

    // only needed in this class
    private String mApiKey;
    private Boolean mIsInDevelopment = false;
    private String mDeveloperPlanId;

    static final int STATUS_DEVICE_CONNECTED = 0;
    static final int STATUS_GATHERING_INFO = 200;
    static final int STATUS_CLEAR_WIFI_COMPLETE = 201;
    static final int STATUS_CLEAR_WIFI_AND_CACHE_COMPLETE = 202;

    static final int ERROR_INVALID_ARGUMENTS = 100;
    static final int ERROR_PROCESS_TIMED_OUT = 101;
    static final int ERROR_CANCELLED_BY_USER = 102;
    static final int ERROR_INVALID_API_KEY = 103;
    static final int ERROR_VERIFY_API_KEY_FAIL = 301; // android only
    static final int ERROR_JSON_ERROR = 302;          // android only

    private static final int START_BLINKUP_ARG_API_KEY = 0;
    private static final int START_BLINKUP_ARG_DEVELOPER_PLAN_ID = 1;
    private static final int START_BLINKUP_IS_IN_DEVELOPMENT = 2;
    private static final int START_BLINKUP_ARG_TIMEOUT_MS = 3;

    /**********************************************************
     * method called by Cordova javascript
     *********************************************************/
    @Override
    public boolean execute(String action, JSONArray data, CallbackContext callbackContext) throws JSONException {
        sCallbackContext = callbackContext;
        final Activity activity = cordova.getActivity();
        final BlinkupController controller = BlinkupController.getInstance();

        if (START_BLINKUP.equalsIgnoreCase(action)) {
            return startBlinkUp(activity, controller, data);
        } else if (ABORT_BLINKUP.equalsIgnoreCase(action)) {
            return abortBlinkup(controller);
        } else if (CLEAR_BLINKUP_DATA.equalsIgnoreCase(action)) {
            return clearBlinkupData(activity, controller);
        }
        return false;
    }

    private boolean startBlinkUp(final Activity activity, final BlinkupController controller, JSONArray data) {
        int timeoutMs;
        try {
            mApiKey = data.getString(START_BLINKUP_ARG_API_KEY);
            mDeveloperPlanId = data.getString(START_BLINKUP_ARG_DEVELOPER_PLAN_ID);
            mIsInDevelopment = data.getBoolean(START_BLINKUP_IS_IN_DEVELOPMENT);
            timeoutMs = data.getInt(START_BLINKUP_ARG_TIMEOUT_MS);
        } catch (JSONException exc) {
            BlinkUpPluginResult.sendPluginErrorToCallback(ERROR_INVALID_ARGUMENTS);
            return false;
        }

        // if api key not valid, send error message and quit
        if (!apiKeyFormatValid()) {
            BlinkUpPluginResult.sendPluginErrorToCallback(ERROR_INVALID_API_KEY);
            return false;
        } else if( mDeveloperPlanId == null || mDeveloperPlanId.isEmpty()) {
            BlinkUpPluginResult.sendPluginErrorToCallback(ERROR_INVALID_ARGUMENTS);
            return false;
        }

        controller.intentBlinkupComplete = createBlinkUpCompleteIntent(activity, timeoutMs);

        // default is to run on WebCore thread, we have UI so need UI thread
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                presentBlinkUp(activity, controller);
            }
        });
        return true;
    }

    private Intent createBlinkUpCompleteIntent(Activity activity, int timeoutMs) {
        Intent blinkupCompleteIntent = new Intent(activity, BlinkUpCompleteActivity.class);
        blinkupCompleteIntent.putExtra(Extras.EXTRA_DEVELOPER_PLAN_ID, mDeveloperPlanId);
        blinkupCompleteIntent.putExtra(Extras.EXTRA_TIMEOUT_MS, timeoutMs);
        return blinkupCompleteIntent;
    }

    private boolean abortBlinkup(BlinkupController controller) {
        controller.cancelTokenStatusPolling();
        BlinkUpPluginResult.sendPluginErrorToCallback(ERROR_CANCELLED_BY_USER);
        return true;
    }

    private boolean clearBlinkupData(final Activity activity, final BlinkupController controller) {
        PreferencesHelper.setPlanId(activity, null);
        sClearCache = true;
        controller.intentClearComplete = new Intent(activity, ClearCompleteActivity.class);

        // default is to run on WebCore thread, clearing shows UI so needs UI thread
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                controller.clearDevice(activity);
            }
        });
        return true;
    }

    /**********************************************************
     * shows BlinkUpPlugin activity and handles appropriate callbacks
     **********************************************************/
    private void presentBlinkUp(Activity activity, BlinkupController controller) {

        // show toast if can't acquire token
        final TokenAcquireCallback tokenAcquireCallback = new TokenAcquireCallback() {
            @Override
            public void onSuccess(String planId, String id) { }

            @Override
            public void onError(String s) {
                Log.e(TAG, s);
            }
        };

        // send back error if connectivity issue
        ServerErrorHandler serverErrorHandler= new ServerErrorHandler() {
            @Override
            public void onError(String s) {
                BlinkUpPluginResult.sendPluginErrorToCallback(ERROR_VERIFY_API_KEY_FAIL);
            }
        };

        // load cached planId if available. Otherwise, SDK generates new one automatically
        // see electricimp.com/docs/manufacturing/planids/ for info about planIDs
        String planId = null;
        if (mIsInDevelopment || org.apache.cordova.BuildConfig.DEBUG) {
            Log.w(TAG, "WARNING - Using Developer Plan. For production, set isInDevelopment flag to false.");
            planId = mDeveloperPlanId;
        } else {
            planId = PreferencesHelper.getPlanId(activity);
        }

        if(planId != null && !planId.isEmpty()) {
            controller.setPlanID(planId);
        }

        controller.acquireSetupToken(activity, mApiKey, tokenAcquireCallback);
        controller.selectWifiAndSetupDevice(activity, mApiKey, serverErrorHandler);
    }

    /**********************************************************
     * @return true if mApiKey is 32 alpha-numeric characters
     *********************************************************/
    private boolean apiKeyFormatValid() {
        if (TextUtils.isEmpty(mApiKey) || TextUtils.getTrimmedLength(mApiKey) != 32) {
            return false;
        }

        String isAlphaNumericPattern = "^[a-zA-Z0-9]*$";
        return mApiKey.matches(isAlphaNumericPattern);
    }

    static boolean getClearCache() {
        return sClearCache;
    }

    static void setClearCache(boolean val) {
        sClearCache = val;
    }

    static CallbackContext getCallbackContext() {
        return sCallbackContext;
    }
}
